File: jira_db_sync/jira_sync/transform.py
"""データ変換

責務:
- Jira API レスポンスから issues_df / change_history_df への変換
- フィールドの抽出・正規化
"""

import json
import logging

import pandas as pd

logger = logging.getLogger(__name__)

# issues テーブルのカラム順（DDL と一致させる）
ISSUES_TABLE_COLUMNS = [
    "id", "project_id", "key", "summary", "description",
    "status", "priority", "assignee", "reporter", "issue_type",
    "resolution", "labels", "components", "fix_versions", "sprint",
    "parent_key", "due_date", "created_date", "updated_date", "raw_data",
]


def _safe_json(val) -> str | None:
    if val is None:
        return None
    return json.dumps(val, ensure_ascii=False)


def _extract_sprint(fields: dict) -> str | None:
    """Sprint フィールドの抽出"""
    for fid in ["sprint", "customfield_10020", "customfield_10104", "customfield_10000"]:
        val = fields.get(fid)
        if val is None:
            continue
        if isinstance(val, list):
            for sp in reversed(val):
                if isinstance(sp, dict) and sp.get("name"):
                    state = sp.get("state", "")
                    if state in ("active", "closed", ""):
                        return sp["name"]
            for sp in val:
                if isinstance(sp, dict) and sp.get("name"):
                    return sp["name"]
        elif isinstance(val, str) and "name=" in val:
            start = val.find("name=") + 5
            end = val.find(",", start)
            if end == -1:
                end = val.find("]", start)
            if end != -1:
                return val[start:end]
    return None


class IssueTransformer:
    """Jira API レスポンスを issues_df に変換"""

    @staticmethod
    def transform(raw_issues: list[dict]) -> pd.DataFrame:
        """raw_issues → DataFrame（テーブルカラム順に並べ替え済み）"""
        rows = []
        for issue in raw_issues:
            f = issue.get("fields", {})
            rows.append({
                "id": issue["id"],
                "key": issue["key"],
                "project_id": f.get("project", {}).get("id", ""),
                "summary": f.get("summary", ""),
                "description": f.get("description"),
                "status": (f.get("status") or {}).get("name"),
                "priority": (f.get("priority") or {}).get("name"),
                "assignee": (f.get("assignee") or {}).get("displayName"),
                "reporter": (f.get("reporter") or {}).get("displayName"),
                "issue_type": (f.get("issuetype") or {}).get("name"),
                "resolution": (f.get("resolution") or {}).get("name"),
                "labels": _safe_json(f.get("labels")),
                "components": _safe_json(
                    [c["name"] for c in (f.get("components") or []) if "name" in c]
                ),
                "fix_versions": _safe_json(
                    [v["name"] for v in (f.get("fixVersions") or []) if "name" in v]
                ),
                "sprint": _extract_sprint(f),
                "parent_key": (f.get("parent") or {}).get("key"),
                "due_date": f.get("duedate"),
                "created_date": f.get("created"),
                "updated_date": f.get("updated"),
                "raw_data": json.dumps(issue, ensure_ascii=False),
            })
        df = pd.DataFrame(rows)
        if not df.empty:
            df = df[ISSUES_TABLE_COLUMNS]
        logger.info("Transformed %d issues", len(df))
        return df


class ChangeHistoryTransformer:
    """Jira API レスポンスから変更履歴を抽出"""

    @staticmethod
    def transform(raw_issues: list[dict]) -> pd.DataFrame:
        """raw_issues → change_history DataFrame"""
        rows = []
        for issue in raw_issues:
            changelog = issue.get("changelog", {})
            for history in changelog.get("histories", []):
                for item in history.get("items", []):
                    rows.append({
                        "issue_id": issue["id"],
                        "issue_key": issue["key"],
                        "history_id": history.get("id", ""),
                        "author_account_id": history.get("author", {}).get("accountId"),
                        "author_display_name": history.get("author", {}).get("displayName"),
                        "field": item.get("field", ""),
                        "field_type": item.get("fieldtype"),
                        "from_value": item.get("from"),
                        "from_string": item.get("fromString"),
                        "to_value": item.get("to"),
                        "to_string": item.get("toString"),
                        "changed_at": history.get("created", ""),
                    })
        df = pd.DataFrame(rows)
        logger.info("Extracted %d change history records", len(df))
        return df

File: jd-go/core/transform.go
package core

import (
	"encoding/json"
	"log/slog"
	"strings"
)

// IssuesTableColumns defines the column order matching DDL.
var IssuesTableColumns = []string{
	"id", "project_id", "key", "summary", "description",
	"status", "priority", "assignee", "reporter", "issue_type",
	"resolution", "labels", "components", "fix_versions", "sprint",
	"parent_key", "due_date", "created_date", "updated_date", "raw_data",
}

func strPtr(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

func safeJSON(val interface{}) *string {
	if val == nil {
		return nil
	}
	b, err := json.Marshal(val)
	if err != nil {
		return nil
	}
	s := string(b)
	return &s
}

func getNestedString(m map[string]interface{}, keys ...string) *string {
	var current interface{} = m
	for _, k := range keys {
		cm, ok := current.(map[string]interface{})
		if !ok || cm == nil {
			return nil
		}
		current = cm[k]
	}
	s, ok := current.(string)
	if !ok || s == "" {
		return nil
	}
	return &s
}

func extractSprint(fields map[string]interface{}) *string {
	fieldIDs := []string{"sprint", "customfield_10020", "customfield_10104", "customfield_10000"}
	for _, fid := range fieldIDs {
		val := fields[fid]
		if val == nil {
			continue
		}
		if arr, ok := val.([]interface{}); ok {
			// First pass: prefer active/closed sprints (reversed)
			for i := len(arr) - 1; i >= 0; i-- {
				sp, ok := arr[i].(map[string]interface{})
				if !ok {
					continue
				}
				name, _ := sp["name"].(string)
				if name == "" {
					continue
				}
				state, _ := sp["state"].(string)
				if state == "active" || state == "closed" || state == "" {
					return &name
				}
			}
			// Second pass: any sprint with a name
			for _, item := range arr {
				sp, ok := item.(map[string]interface{})
				if !ok {
					continue
				}
				name, _ := sp["name"].(string)
				if name != "" {
					return &name
				}
			}
		} else if s, ok := val.(string); ok && strings.Contains(s, "name=") {
			start := strings.Index(s, "name=") + 5
			end := strings.Index(s[start:], ",")
			if end == -1 {
				end = strings.Index(s[start:], "]")
			}
			if end != -1 {
				name := s[start : start+end]
				return &name
			}
		}
	}
	return nil
}

// TransformIssues converts raw Jira API responses to Issue structs.
func TransformIssues(rawIssues []map[string]interface{}) []Issue {
	issues := make([]Issue, 0, len(rawIssues))
	for _, issue := range rawIssues {
		f, _ := issue["fields"].(map[string]interface{})
		if f == nil {
			f = map[string]interface{}{}
		}

		// Extract components names
		var componentNames []string
		if comps, ok := f["components"].([]interface{}); ok {
			for _, c := range comps {
				if cm, ok := c.(map[string]interface{}); ok {
					if name, ok := cm["name"].(string); ok {
						componentNames = append(componentNames, name)
					}
				}
			}
		}

		// Extract fix version names
		var fixVersionNames []string
		if versions, ok := f["fixVersions"].([]interface{}); ok {
			for _, v := range versions {
				if vm, ok := v.(map[string]interface{}); ok {
					if name, ok := vm["name"].(string); ok {
						fixVersionNames = append(fixVersionNames, name)
					}
				}
			}
		}

		// Description may be a complex object in API v3
		var desc *string
		if d := f["description"]; d != nil {
			switch v := d.(type) {
			case string:
				desc = &v
			default:
				b, _ := json.Marshal(v)
				s := string(b)
				desc = &s
			}
		}

		rawJSON, _ := json.Marshal(issue)

		projectID := ""
		if proj, ok := f["project"].(map[string]interface{}); ok {
			projectID, _ = proj["id"].(string)
		}

		iss := Issue{
			ID:          issue["id"].(string),
			Key:         issue["key"].(string),
			ProjectID:   projectID,
			Summary:     getString(f, "summary"),
			Description: desc,
			Status:      getNestedString(f, "status", "name"),
			Priority:    getNestedString(f, "priority", "name"),
			Assignee:    getNestedString(f, "assignee", "displayName"),
			Reporter:    getNestedString(f, "reporter", "displayName"),
			IssueType:   getNestedString(f, "issuetype", "name"),
			Resolution:  getNestedString(f, "resolution", "name"),
			Labels:      safeJSON(f["labels"]),
			Components:  safeJSON(componentNames),
			FixVersions: safeJSON(fixVersionNames),
			Sprint:      extractSprint(f),
			ParentKey:   getNestedString(f, "parent", "key"),
			DueDate:     getStringPtr(f, "duedate"),
			CreatedDate: getStringPtr(f, "created"),
			UpdatedDate: getStringPtr(f, "updated"),
			RawData:     string(rawJSON),
		}
		issues = append(issues, iss)
	}
	slog.Info("Transformed issues", "count", len(issues))
	return issues
}

func getString(m map[string]interface{}, key string) string {
	s, _ := m[key].(string)
	return s
}

func getStringPtr(m map[string]interface{}, key string) *string {
	v := m[key]
	if v == nil {
		return nil
	}
	s, ok := v.(string)
	if !ok || s == "" {
		return nil
	}
	return &s
}

// TransformChangeHistory extracts change history from raw issues.
func TransformChangeHistory(rawIssues []map[string]interface{}) []ChangeHistory {
	var rows []ChangeHistory
	for _, issue := range rawIssues {
		issueID, _ := issue["id"].(string)
		issueKey, _ := issue["key"].(string)

		changelog, _ := issue["changelog"].(map[string]interface{})
		histories, _ := changelog["histories"].([]interface{})

		for _, h := range histories {
			hist, ok := h.(map[string]interface{})
			if !ok {
				continue
			}
			historyID, _ := hist["id"].(string)
			changedAt, _ := hist["created"].(string)
			var authorAccountID, authorDisplayName *string
			if author, ok := hist["author"].(map[string]interface{}); ok {
				authorAccountID = getStringPtr(author, "accountId")
				authorDisplayName = getStringPtr(author, "displayName")
			}

			items, _ := hist["items"].([]interface{})
			for _, it := range items {
				item, ok := it.(map[string]interface{})
				if !ok {
					continue
				}
				field, _ := item["field"].(string)
				rows = append(rows, ChangeHistory{
					IssueID:           issueID,
					IssueKey:          issueKey,
					HistoryID:         historyID,
					AuthorAccountID:   authorAccountID,
					AuthorDisplayName: authorDisplayName,
					Field:             field,
					FieldType:         getStringPtr(item, "fieldtype"),
					FromValue:         getStringPtr(item, "from"),
					FromString:        getStringPtr(item, "fromString"),
					ToValue:           getStringPtr(item, "to"),
					ToString:          getStringPtr(item, "toString"),
					ChangedAt:         changedAt,
				})
			}
		}
	}
	slog.Info("Extracted change history records", "count", len(rows))
	return rows
}
